lib/os/sem.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <misc/sem.h>
+#include <syscall_handler.h>
+
+#ifdef CONFIG_USERSPACE
+int sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
+		unsigned int limit)
+{
+	if (sem == NULL || limit == 0 ||
+	    initial_count > limit || limit > INT_MAX) {
+		return -EINVAL;
+	}
+
+	atomic_set(&sem->futex.val, initial_count);
+	sem->limit = limit;
+
+	return 0;
+}
+
+unsigned int sys_sem_count_get(struct sys_sem *sem)
+{
+	int value = atomic_get(&sem->futex.val);
+
+	return value > 0 ? value : 0;
+}
+
+int sys_sem_give(struct sys_sem *sem)
+{
+	int ret = 0;
+	int old_value;
+
+	do {
+		old_value = atomic_get(&sem->futex.val);
+		if (old_value >= sem->limit) {
+			break;
+		}
+	} while (atomic_cas(&sem->futex.val, old_value,
+			old_value < 0 ? 1 : old_value + 1) == 0);
+
+	if (old_value < 0) {
+		ret = k_futex_wake(&sem->futex, true);
```

Comment:
Why true here for the wake_all parameter?  You've incremented the semaphore by one, we only need to wake up one thread, by definition.

Reply:
@andyross I use -1(futex.val) to donate there is contend on the sys_sem, maybe there is only one waiter or many waiters on this sys_sem. For many waiters case, if just wake up one waiter, others have no chance to be woken up. And wake them all, it will sleep wait again in sys_sem_take. And for one waiter case, wake all == wake one.

For why I use -1 instead of just decrement, because when decrement to -INT_MAX, we also should consider the left bound.

Reply:
I think this is fine, BIONIC does the same thing.